homeassistant/components/shelly/utils.py

```diff
@@ -567,14 +567,15 @@ def async_remove_orphaned_virtual_entities(
     for entity in entities:
         if not entity.entity_id.startswith(platform):
             continue
-        if virt_comp_type not in entity.unique_id:
+        if key_suffix is not None and key_suffix not in entity.unique_id:
             continue
-        # we are looking for the component ID, e.g. boolean:201
-        if not (match := re.search(r"[a-z]+:\d+", entity.unique_id)):
+        # we are looking for the component ID, e.g. boolean:201, em1data:1
+        if not (match := re.search(r"[a-z\d]+:\d+", entity.unique_id)):
```

Comment:
The re could be pre-compiled as a constant instead of compiling every loop since it never changes